docs/design/interoperability/README.md

```diff
@@ -29,4 +50,209 @@ more detail.
 
 ## Overview
 
-TODO
+Carbon's bidirectional interoperability with C++ is
+[a cornerstone of its design](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code),
+enabling a gradual transition from existing C++ codebases. The goal is not just
+a foreign function interface (FFI), but a seamless, high-fidelity integration
+that supports advanced C++ features, from templates to class hierarchies.
+
+C++ APIs are imported into Carbon using an `import Cpp` directive, which makes
+C++ declarations available within a dedicated `Cpp` namespace in Carbon. This
+prevents name collisions and makes the origin of symbols explicit. Carbon code
+can then call C++ functions, instantiate C++ classes, and use C++ types, while
+respecting C++'s semantics, including its complex overload resolution rules.
+
+Similarly, Carbon APIs can be designed to be callable from C++. The
+interoperability layer is designed to be zero-cost, avoiding unnecessary
+allocations or copies when calling between the two languages This is achieved
+through a deep semantic co-design, where the Carbon compiler embeds a C++
+compiler frontend (Clang) to understand and map C++ constructs with high
+fidelity. This includes preserving the nominal distinctions between C++ types
+like `long` and `long long`, or `T*` and `T&`, which is critical for correct
+overload resolution and template instantiation.
+
+## C++ Interoperability Model: Introduction and Principles
+
+### The "Successor Language" Mandate
+
+The design of Carbon's C++ interoperability is governed by its foundational
+goal: [to be a successor language](/README.md), not merely a language with a
+foreign function interface (FFI). This mandate dictates a design that moves
+beyond the C-style FFI adopted by most modern languages and instead provides
+"seamless, bidirectional interoperability". The objective is to support deep
+integration with existing C++ code, encompassing its most complex features,
+"from inheritance to templates".
+
+This goal has profound implications for the Carbon compiler and language
+semantics. It requires that C++ is not treated as a "foreign" entity. Instead,
+Carbon's semantic model must be _co-designed_ to understand, map, and interact
+with C++'s semantic constructs—including templates, class hierarchies, and
+complex overload resolution—with high fidelity. The interoperability layer must,
+therefore, operate at the semantic analysis (SemIR) level, not just at the
+linking (ABI) level. This document specifies the design of this semantic
+contract.
+
+### The `Cpp` Associated Type: The Interop Trigger
+
+A core mechanism in this design is the `Cpp` associated type. This concept
```

Comment:
Isn't `Cpp` a package namespace, not an associated type? What led you to the "associated type" terminology here?

Carbon's associated facet is called an associated type in other languages (Rust and Swift in particular, as noted at https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#associated-facets). Perhaps something like "C++ interop type" might avoid terminology overlap?